File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/AnalyticsEventsSubmissionService.java
package uk.nhs.nhsx.analyticsevents;

import org.apache.http.entity.ContentType;
import uk.nhs.nhsx.analyticssubmission.PostDistrictLaReplacer;
import uk.nhs.nhsx.analyticssubmission.model.PostDistrictLADTuple;
import uk.nhs.nhsx.core.Jackson;
import uk.nhs.nhsx.core.aws.s3.BucketName;
import uk.nhs.nhsx.core.aws.s3.ByteArraySource;
import uk.nhs.nhsx.core.aws.s3.Locator;
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider;
import uk.nhs.nhsx.core.aws.s3.S3Storage;
import uk.nhs.nhsx.core.events.Events;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

public class AnalyticsEventsSubmissionService {

    private final S3Storage s3Storage;
    private final ObjectKeyNameProvider objectKeyNameProvider;
    private final BucketName bucketName;
    private final Events events;

    public AnalyticsEventsSubmissionService(S3Storage s3Storage, ObjectKeyNameProvider objectKeyNameProvider, BucketName bucketName, Events events) {
        this.s3Storage = s3Storage;
        this.objectKeyNameProvider = objectKeyNameProvider;
        this.bucketName = bucketName;
        this.events = events;
    }

    public void accept(Map<String, Object> payload) {
        var storedPayload = transformPayload(payload);
        String json = Jackson.toJson(storedPayload);
        uploadToS3(json);
    }

    private Map<String, Object> transformPayload(Map<String, Object> payload) {
        Object metadataRaw = payload.get("metadata");

        if (!(metadataRaw instanceof Map)) {
            throw new RuntimeException("metadata must be a map");
        }

        var metadata = (Map<?, ?>) metadataRaw;
        var currentPostalDistrict = (String) metadata.get("postalDistrict");

        var transformedMetadata = new LinkedHashMap<Object, Object>(metadata);
        currentPostalDistrict = (String) metadata.get("postalDistrict");
        var currentLocalAuthority = (String) metadata.get("localAuthority");
        PostDistrictLADTuple newPostDistrictLATuple = PostDistrictLaReplacer.replacePostDistrictLA(currentPostalDistrict, currentLocalAuthority, events);

        transformedMetadata.put("postalDistrict", newPostDistrictLATuple.postDistrict);
        transformedMetadata.put("localAuthority", newPostDistrictLATuple.localAuthorityId);

        var transformedPayload = new LinkedHashMap<>(payload);
        transformedPayload.put("uuid", UUID.randomUUID());
        transformedPayload.put("metadata", transformedMetadata);

        return transformedPayload;
    }

    private void uploadToS3(String json) {
        var objectKey = objectKeyNameProvider.generateObjectKeyName().append(".json");

        s3Storage.upload(
            Locator.of(bucketName, objectKey),
            ContentType.APPLICATION_JSON,
            ByteArraySource.fromUtf8String(json)
        );
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticsevents/AnalyticsEventsSubmissionService.kt
package uk.nhs.nhsx.analyticsevents

import uk.nhs.nhsx.analyticssubmission.PostDistrictLaReplacer.replacePostDistrictLA
import uk.nhs.nhsx.core.ContentType.Companion.APPLICATION_JSON
import uk.nhs.nhsx.core.Jackson.toJson
import uk.nhs.nhsx.core.aws.s3.BucketName
import uk.nhs.nhsx.core.aws.s3.ByteArraySource.Companion.fromUtf8String
import uk.nhs.nhsx.core.aws.s3.Locator
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider
import uk.nhs.nhsx.core.aws.s3.S3Storage
import uk.nhs.nhsx.core.events.Events
import java.util.*

class AnalyticsEventsSubmissionService(
    private val s3Storage: S3Storage,
    private val objectKeyNameProvider: ObjectKeyNameProvider,
    private val bucketName: BucketName,
    private val events: Events
) {
    fun accept(payload: Map<String, Any>) {
        uploadToS3(toJson(transformPayload(payload)))
    }

    private fun transformPayload(payload: Map<String, Any>): Map<String, Any> {
        val metadata = payload["metadata"] as? Map<*, *> ?: error("metadata must be a map")

        val currentPostalDistrict = metadata["postalDistrict"] as String?
        val currentLocalAuthority = metadata["localAuthority"] as String?
        val mappedPostDistrict = replacePostDistrictLA(
            currentPostalDistrict,
            currentLocalAuthority,
            events
        )

        return payload.toMutableMap().apply {
            this["uuid"] = UUID.randomUUID()
            this["metadata"] = metadata.toMutableMap().apply {
                this["postalDistrict"] = mappedPostDistrict.postDistrict
                this["localAuthority"] = mappedPostDistrict.localAuthorityId
            }

        }
    }

    private fun uploadToS3(json: String) {
        s3Storage.upload(
            Locator.of(bucketName, objectKeyNameProvider.generateObjectKeyName().append(".json")),
            APPLICATION_JSON,
            fromUtf8String(json)
        )
    }
}
